### PR TITLE
feat(logs): support data protection custom data identifiers

### LINKED ISSUE #28430: aws-logs: support custom data identfiers
### Describe the feature

Support the newly launched custom data identifiers feature for CloudWatch Logs sensitive data protection. 

https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html

### Use Case

Custom data identifiers (CDIs) let you define your own custom regular expressions that can be used in your data protection policy. Using custom data identifiers, you can target business-specific personally identifiable information (PII) use cases that managed data identifiers can't provide. For example, you can use a custom data identifier to look for company-specific employee IDs. Custom data identifiers can be used in conjunction with managed data identifiers.

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

Latest

### Environment details (OS name and version, etc.)

All

### PR DESCRIPTION
**Pull request by kchg**
### Feature

Support the newly launched custom data identifiers feature for CloudWatch Logs sensitive data protection.

https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html

### Use Case

Custom data identifiers (CDIs) let you define your own custom regular expressions that can be used in your data protection policy. Using custom data identifiers, you can target business-specific personally identifiable information (PII) use cases that managed data identifiers can't provide. For example, you can use a custom data identifier to look for company-specific employee IDs. Custom data identifiers can be used in conjunction with managed data identifiers.

### Solution
Users can now supply a `regex` field to the `DataIdentifiers` constructor. Supplying this field will enable the named identifier as a custom data identifier. 

Closes #28430.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks 👍
I left a couple of suggestions for changes in the implementation.

File: packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts [hunk 1]
```diff
@@ -274,9 +299,9 @@ export class DataIdentifier {
   public static readonly VEHICLEIDENTIFICATIONNUMBER = new DataIdentifier('VehicleIdentificationNumber');
   public static readonly ZIPCODE_US = new DataIdentifier('ZipCode-US');
 
-  constructor(private readonly identifier: string) { }
+  constructor(public readonly name: string, public readonly regex?: string) { }
```
[the remark this answers is not in the document]
Comment (kchg) [reply packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts]: Done, thanks for the suggestion.

File: packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts [hunk 2]
```diff
@@ -144,7 +169,9 @@ export interface DataProtectionPolicyProps {
   readonly description?: string;
 
   /**
-   * List of data protection identifiers. Must be in the following list: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/protect-sensitive-log-data-types.html
+   * List of data protection identifiers, containing managed or custom data identfiers (CustomDataIdentifier).
+   * Managed data identiers must be in the following list: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-managed-data-identifiers.html
+   * Custom data identfiers must have a valid regex defined: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html
    *
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts:175]:
```suggestion
   * List of data protection identifiers.
   *
   * Managed data identifiers must be in the following list: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-managed-data-identifiers.html
   * Custom data identifiers must have a valid regex defined: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html#custom-data-identifiers-constraints
```
Formatting and typo.

File: packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts [hunk 3]
```diff
@@ -274,9 +301,36 @@ export class DataIdentifier {
...
 
   public toString(): string {
-    return this.identifier;
+    return this.name;
+  }
+}
+
+/**
+ * A custom data identifier. Include a custom data identifier name and regular expression in the JSON policy used to define the data protection policy.
+ * https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL-custom-data-identifiers.html
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts:317]:
```suggestion
```
We already provided the link multiple times.

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks for the changes 👍
I left some notes for some minor adjustments.

File: packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts [hunk 4]
```diff
@@ -274,9 +303,35 @@ export class DataIdentifier {
...
+    return this.name;
+  }
+}
+
+/**
+ * A custom data identifier. Include a custom data identifier name and regular expression in the JSON policy used to define the data protection policy.
+ */
+export class CustomDataIdentifier extends DataIdentifier {
+  /**
+   * Create a custom data identfier
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-logs/lib/data-protection-policy.ts:322]:
```suggestion
   * Create a custom data identifier.
```
Typo

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍 (please fix the typo in the documentation)

**Review by paulhcsun — CHANGES_REQUESTED** [review]
Overall looks good to me, just one merge conflict that needs to be resolved @kchg.

Thanks for reviewing @lpizzinidev!

**Review by paulhcsun — DISMISSED** [review]
Overall looks good to me, just one merge conflict that needs to be resolved @kchg.

Thanks for reviewing @lpizzinidev!

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**